src/bisimulation.rs

```diff
@@ -0,0 +1,372 @@
+
+        let leaf_node_block = FineBlock {
+            values: leaf_node_indices,
+            coarse_block_that_supersets_self: Rc::clone(&coarse_initial_block_pointer),
+        };
+        let non_leaf_node_block = FineBlock {
+            values: non_leaf_node_indices,
+            coarse_block_that_supersets_self: Rc::clone(&coarse_initial_block_pointer),
+        };
+
+        (
+            Rc::new(RefCell::new(leaf_node_block)),
+            Rc::new(RefCell::new(non_leaf_node_block)),
+        )
+    };
+
+    coarse_initial_block_pointer
+        .fine_blocks_that_are_subsets_of_self
+        .borrow_mut()
+        .extend([
+            Rc::clone(&leaf_node_block_pointer),
+            Rc::clone(&non_leaf_node_block_pointer),
+        ]);
+
+    let node_to_block_vec = (*leaf_node_block_pointer)
+        .borrow()
+        .values
+        .clone()
+        .iter()
+        .fold(
+            vec![
+                Rc::clone(&non_leaf_node_block_pointer);
+                (*graph_node_indices.last().unwrap()).index() as usize + 1
+            ],
+            |mut accumulator, value| {
+                accumulator[value.index()] = Rc::clone(&leaf_node_block_pointer);
+                accumulator
+            },
+        );
+
+    (
+        (leaf_node_block_pointer, non_leaf_node_block_pointer),
+        vec![coarse_initial_block_pointer],
+        node_to_block_vec,
+    )
+}
+
+fn build_counterimage<IndexHolder: HasValues>(
+    graph: &StablePyGraph<Directed>,
+    fine_block: IndexHolder,
+) -> Counterimage {
+    let mut counterimage = HashMap::new();
+    fine_block.values().iter().for_each(|node_index_pointer| {
+        counterimage.insert(
+            *node_index_pointer,
+            graph
+                .neighbors_directed(*node_index_pointer, Incoming)
+                .collect::<Vec<graph::NodeIndex>>(),
+        );
+    });
+
+    counterimage
+}
+
+fn group_by_counterimage(
+    counterimage: Counterimage,
+    node_to_block: &NodeToBlockVec,
+) -> CounterimageGrouped {
+    let mut counterimage_grouped: CounterimageGrouped = HashMap::new();
+
+    for incoming_neighbor_group in counterimage.values() {
+        for node in incoming_neighbor_group {
+            let block = Rc::clone(&node_to_block[node.index()]);
+
+            let key = (*block).borrow().values.clone();
+            match counterimage_grouped.contains_key(&key) {
+                true => {
+                    counterimage_grouped
+                        .get_mut(&key)
+                        .unwrap()
+                        .subblock
+                        .push(*node);
+                }
+                false => {
+                    counterimage_grouped.insert(
+                        key,
+                        CounterImageGroup {
+                            block: Rc::clone(&block),
+                            subblock: Vec::from([*node]),
+                        },
```

Comment:
Minor Rust comment: it is more idiomatic to write [`.entry()` and then handle the Occupied/Vacant case](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html). It would also save a hashmap lookup and be slightly faster